File: src/lexer.cpp
#include "include/lexer.hpp"
#include "errors.hpp"

std::string getStringOnLineNumber(const std::string &str, int lineNumber)
{
    size_t startPos = 0;
    for (int i = 1; i < lineNumber; ++i)
    {
        if ((startPos = str.find('\n', startPos)) == std::string::npos)
        {
            return "";
        }
        startPos++;
    }
    size_t endPos = str.find('\n', startPos);
    return str.substr(startPos, (endPos != std::string::npos) ? endPos - startPos : std::string::npos);
}

Lexer::Lexer(std::string source)
{
    this->source = source;

    this->pos = -1;
    this->lineNo = 1;
    this->colNo = -1;
    this->currentChar = "\0";
    this->_readChar();
}

void Lexer::_readChar()
{
    this->pos++;
    if (this->pos >= this->source.length())
    {
        this->currentChar = "\0";
    }
    else
    {
        this->currentChar = this->source[this->pos];
    }
    this->colNo++;
};

std::shared_ptr<std::string> Lexer::_peekChar(int offset)
{
    if ((this->pos + offset) >= this->source.length())
    {
        return std::make_shared<std::string>("\0");
    }
    else
    {
        return std::make_shared<std::string>(1, this->source[this->pos + offset]);
    }
}

token::tokenType Lexer::_lookupIdent(std::shared_ptr<std::string> ident)
{
    if (*ident == "and")
    {
        return token::tokenType::And;
    }
    else if (*ident == "or")
    {
        return token::tokenType::Or;
    }
    else if (*ident == "not")
    {
        return token::tokenType::Not;
    }
    else if (*ident == "var")
    {
        return token::tokenType::Var;
    }
    else if (*ident == "def")
    {
        return token::tokenType::Def;
    }
    else if (*ident == "return")
    {
        return token::tokenType::Return;
    }
    else if (*ident == "if")
    {
        return token::tokenType::If;
    }
    else if (*ident == "else")
    {
        return token::tokenType::Else;
    }
    else if (*ident == "elif")
    {
        return token::tokenType::ElIf;
    }
    else if (*ident == "is")
    {
        return token::tokenType::Is;
    }
    else if (*ident == "while")
    {
        return token::tokenType::While;
    }
    else if (*ident == "for")
    {
        return token::tokenType::For;
    }
    else if (*ident == "in")
    {
        return token::tokenType::In;
    }
    else if (*ident == "break")
    {
        return token::tokenType::Break;
    }
    else if (*ident == "continue")
    {
        return token::tokenType::Continue;
    }
    else if (*ident == "True")
    {
        return token::tokenType::True;
    }
    else if (*ident == "False")
    {
        return token::tokenType::False;
    }
    else if (*ident == "MayBe")
    {
        return token::tokenType::Maybe;
    }
    else if (*ident == "None")
    {
        return token::tokenType::None;
    }
    return token::tokenType::Identifier;
};

std::shared_ptr<token::Token> Lexer::nextToken()
{
    std::shared_ptr<token::Token> token;

    this->_skipWhitespace();

    if (this->currentChar == "+")
    {
        if (*this->_peekChar() == "+")
        {
            token = this->_newToken(token::tokenType::Increment, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::PlusEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Plus, this->currentChar);
        }
    }
    else if (this->currentChar == "-")
    {
        if (*this->_peekChar() == ">")
        {
            token = this->_newToken(token::tokenType::RightArrow, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (*this->_peekChar() == "-")
        {
            token = this->_newToken(token::tokenType::Decrement, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (this->_isDigit(*(this->_peekChar())))
        {
            this->_readChar();
            token = this->_readNumber();
            token->literal = "-" + token->literal;
        }
        else if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::DashEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Dash, this->currentChar);
        }
    }
    else if (this->currentChar == "*")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::AsteriskEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (*this->_peekChar() == "*")
        {
            token = this->_newToken(token::tokenType::AsteriskAsterisk, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Asterisk, this->currentChar);
        }
    }
    else if (this->currentChar == "/")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::ForwardSlashEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::ForwardSlash, this->currentChar);
        }
    }
    else if (this->currentChar == "%")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::PercentEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Percent, this->currentChar);
        }
    }
    else if (this->currentChar == "^")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::CaretEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::BitwiseXor, this->currentChar);
        }
    }
    else if (this->currentChar == "=")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::EqualEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Equals, this->currentChar);
        }
    }
    else if (this->currentChar == ">")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::GreaterThanOrEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (*this->_peekChar() == ">")
        {
            token = this->_newToken(token::tokenType::RightShift, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::GreaterThan, this->currentChar);
        }
    }
    else if (this->currentChar == "<")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::LessThanOrEqual, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else if (*this->_peekChar() == "<")
        {
            token = this->_newToken(token::tokenType::LeftShift, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::LessThan, this->currentChar);
        }
    }
    else if (this->currentChar == "!")
    {
        if (*this->_peekChar() == "=")
        {
            token = this->_newToken(token::tokenType::NotEquals, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Illegal, this->currentChar);
        }
    }
    else if (this->currentChar == "{")
    {
        token = this->_newToken(token::tokenType::LeftBrace, this->currentChar);
    }
    else if (this->currentChar == "}")
    {
        token = this->_newToken(token::tokenType::RightBrace, this->currentChar);
    }
    else if (this->currentChar == "(")
    {
        token = this->_newToken(token::tokenType::LeftParen, this->currentChar);
    }
    else if (this->currentChar == ")")
    {
        token = this->_newToken(token::tokenType::RightParen, this->currentChar);
    }
    else if (this->currentChar == "{")
    {
        token = this->_newToken(token::tokenType::LeftBrace, this->currentChar);
    }
    else if (this->currentChar == "}")
    {
        token = this->_newToken(token::tokenType::RightBrace, this->currentChar);
    }
    else if (this->currentChar == ":")
    {
        token = this->_newToken(token::tokenType::Colon, this->currentChar);
    }
    else if (this->currentChar == ";")
    {
        token = this->_newToken(token::tokenType::Semicolon, this->currentChar);
    }
    else if (this->currentChar == "&")
    {
        if (*this->_peekChar() == "&")
        {
            token = this->_newToken(token::tokenType::BitwiseAnd, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Illegal, this->currentChar);
        }
    }
    else if (this->currentChar == "|")
    {
        if (*this->_peekChar() == "|")
        {
            token = this->_newToken(token::tokenType::BitwiseOr, this->currentChar + *this->_peekChar());
            this->_readChar();
        }
        else
        {
            token = this->_newToken(token::tokenType::Illegal, this->currentChar);
        }
    }
    else if (this->currentChar == "~")
    {
        token = this->_newToken(token::tokenType::BitwiseNot, this->currentChar);
    }
    else if (this->currentChar == ",")
    {
        token = this->_newToken(token::tokenType::Comma, this->currentChar);
    }
    else if (this->currentChar == "\0")
    {
        token = this->_newToken(token::tokenType::EndOfFile, "");
    }
    else
    {
        if (this->currentChar == "\"")
        {
            std::shared_ptr<std::string> str = this->_readString("\"");
            token = this->_newToken(token::tokenType::String, *str);
            return token;
        }
        else if (this->_isLetter(this->currentChar))
        {
            std::shared_ptr<std::string> ident = this->_readIdentifier();
            token = this->_newToken(this->_lookupIdent(ident), *ident);
            return token;
        }
        else if (this->_isDigit(this->currentChar))
        {
            token = this->_readNumber();
            return token;
        }
        else
        {
            token = this->_newToken(token::tokenType::Illegal, this->currentChar);
        }
    }
    this->_readChar();
    return token;
}

std::shared_ptr<token::Token> Lexer::_newToken(token::tokenType type, std::string currentChar)
{
    auto x = std::make_shared<token::Token>(type, currentChar, this->lineNo, this->colNo);
    return x;
}

std::shared_ptr<token::Token> Lexer::_readNumber()
{
    int dot_count = 0;
    std::string number = "";
    while (this->_isDigit(this->currentChar) || this->currentChar == ".")
    {
        if (this->currentChar == ".")
        {
            dot_count++;
            if (dot_count > 1)
            {
                printf("Invalid number at line %d, column %d\n", this->lineNo, this->colNo);
                return this->_newToken(token::tokenType::Illegal, this->currentChar);
            }
        }
        number += this->currentChar;
        this->_readChar();
        if (this->currentChar == "\0")
            break;
    }
    if (dot_count == 0)
    {
        return this->_newToken(token::tokenType::Integer, number);
    }
    return this->_newToken(token::tokenType::Float, number);
};

std::shared_ptr<std::string> Lexer::_readIdentifier()
{
    std::string identifier = "";
    while (this->_isLetter(this->currentChar) || this->_isDigit(this->currentChar))
    {
        identifier += this->currentChar;
        this->_readChar();
    };
    return std::make_shared<std::string>(identifier);
}

void Lexer::_skipWhitespace()
{
    while (this->currentChar == " " || this->currentChar == "\t" || this->currentChar == "\n" || this->currentChar == "\r")
    {
        if (this->currentChar == "\n")
        {
            this->lineNo++;
            this->colNo = 0;
        }

        this->_readChar();
    }
    if (this->currentChar == "#")
    {
        this->_readChar();
        while (this->currentChar != "\n" && this->currentChar != "\0")
        {
            this->_readChar();
        }
        this->_skipWhitespace();
    }
}

std::shared_ptr<std::string> Lexer::_readString(std::string quote)
{
    std::string str = "";
    std::string literal = "";
    this->_readChar();
    while (this->currentChar != quote && this->currentChar != "\0")
    {
        if ((quote != "\"\"\"" && quote != "'''") && (this->currentChar == "\n" || this->currentChar == "\r"))
        {
            raiseLexerError("Invalid string because no closing quote found on the same line.", this->source, this->lineNo, this->colNo, literal);
        }
        else if (this->currentChar == "\0")
        {
            raiseLexerError("Invalid string because no closing quote found.", this->source, this->lineNo, this->colNo, literal);
        }
        else if (this->currentChar == "\\")
        {
            this->_readChar();
            if (this->currentChar == "n")
            {
                str += "\n";
                literal += "\\n";
            }
            else if (this->currentChar == "t")
            {
                str += "\t";
                literal += "\\t";
            }
            else if (this->currentChar == "r")
            {
                str += "\r";
                literal += "\\r";
            }
            else if (this->currentChar == "b")
            {
                str += "\b";
                literal += "\\b";
            }
            else if (this->currentChar == "a")
            {
                str += "\a";
                literal += "\\a";
            }
            else if (this->currentChar == "f")
            {
                str += "\f";
                literal += "\\f";
            }
            else if (this->currentChar == "v")
            {
                str += "\v";
                literal += "\\v";
            }
            else if (this->currentChar == "\\")
            {
                str += "\\";
                literal += "\\\\";
            }
            else if (this->currentChar == "\"")
            {
                str += "\"";
                literal += "\\\"";
            }
            else if (this->currentChar == "'")
            {
                str += "'";
                literal += "\\'";
            }
            else
            {
                str += "\\" + this->currentChar;
                literal += "\\" + this->currentChar;
            }
        }
        else
        {
            str += this->currentChar;
            literal += this->currentChar;
        }
        this->_readChar();
    }
    this->_readChar();
    return std::make_shared<std::string>(str);
}

bool Lexer::_isDigit(std::string character)
{
    return character >= "0" && character <= "9";
};

bool Lexer::_isLetter(std::string character)
{
    return (character >= "a" && character <= "z") || (character >= "A" && character <= "Z") || character == "_";
};

File: src/lexer.py
from tokens import Token, TokenType


class Lexer:
    def __init__(self, source: str) -> None:
        self.tokens: list[Token] = []
        self.source = source
        self.pos = 0
        self.current_char: str = ""
        self.line = 1
        self.column = 0
        self._read_char()
        while self.current_char != "":
            self.tokens.append(self.next_token())

    def next_token(self) -> Token:
        tok: Token
        self._skip_blanks()
        match self.current_char:
            case "":
                tok = Token(TokenType.EOF, "", self.line, self.column, self.column)
            case "+":
                if self._peek_char() == "=":
                    tok = Token(TokenType.PlusEqual, "+=", self.line, self.column, self.column + 1)
                    self.pos += 1
                elif self._peek_char() == "+":
                    tok = Token(TokenType.Increment, "++", self.line, self.column, self.column + 1)
                else:
                    tok = Token(TokenType.Plus, "+", self.line, self.column, self.column)
            case "-":
                if self._peek_char() == "=":
                    tok = Token(TokenType.DashEqual, "-=", self.line, self.column, self.column + 1)
                    self.pos += 1
                elif self._peek_char() == "-":
                    tok = Token(TokenType.Decrement, "--", self.line, self.column, self.column + 1)
                elif self._peek_char() == ">":
                    tok = Token(TokenType.RightArrow, "->", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.Dash, "-", self.line, self.column, self.column)
            case "*":
                if self._peek_char() == "*":
                    tok = Token(TokenType.AsteriskAsterisk, "**", self.line, self.column, self.column + 1)
                    self.pos += 1
                elif self._peek_char() == "=":
                    tok = Token(TokenType.AsteriskEqual, "*=", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.Asterisk, "*", self.line, self.column, self.column)
            case "%":
                if self._peek_char() == "=":
                    tok = Token(TokenType.PercentEqual, "%=", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.Percent, "%", self.line, self.column, self.column)
            case "/":
                if self._peek_char() == "=":
                    tok = Token(TokenType.ForwardSlashEqual, "/=", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.ForwardSlash, "/", self.line, self.column, self.column)
            case "\\":
                if self._peek_char() == "=":
                    tok = Token(TokenType.BackwardSlashEqual, "\\=", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.BackwardSlash, "\\", self.line, self.column, self.column)
            case "=":
                if self._peek_char() == "=":
                    tok = Token(TokenType.EQ, "==", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.Equals, "=", self.line, self.column, self.column)
            case ".":
                if self._peek_char() == "." and self._peek_char(2) == ".":
                    tok = Token(TokenType.Ellipsis, "...", self.line, self.column, self.column + 2)
                    self.pos += 2
                else:
                    tok = Token(TokenType.Dot, ".", self.line, self.column, self.column)
            case "!":
                if self._peek_char() == "=":
                    tok = Token(TokenType.NOT_EQ, "!=", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.ILLEGAL, self.current_char, self.line, self.column, self.column)
            case "<":
                if self._peek_char() == "=":
                    tok = Token(TokenType.LTE, "<=", self.line, self.column, self.column + 1)
                    self.pos += 1
                elif self._peek_char() == "<":
                    tok = Token(TokenType.LeftShift, "<<", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.LT, "<", self.line, self.column, self.column)
            case ">":
                if self._peek_char() == "=":
                    tok = Token(TokenType.GTE, ">=", self.line, self.column, self.column + 1)
                    self.pos += 1
                elif self._peek_char() == ">":
                    tok = Token(TokenType.RightShift, ">>", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.GT, ">", self.line, self.column, self.column)
            case "^":
                tok = Token(TokenType.BitwiseXor, "^", self.line, self.column, self.column)
            case "~":
                tok = Token(TokenType.BitwiseNot, "~", self.line, self.column, self.column)
            case "&":
                if self._peek_char() == "&":
                    tok = Token(TokenType.BitwiseAnd, "&&", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.ILLEGAL, self.current_char, self.line, self.column, self.column)
            case "|":
                if self._peek_char() == "|":
                    tok = Token(TokenType.BitwiseOr, "||", self.line, self.column, self.column + 1)
                    self.pos += 1
                else:
                    tok = Token(TokenType.ILLEGAL, self.current_char, self.line, self.column, self.column)
            case "(":
                tok = Token(TokenType.LeftParen, "(", self.line, self.column, self.column)
            case ")":
                tok = Token(TokenType.RightParen, ")", self.line, self.column, self.column)
            case "{":
                tok = Token(TokenType.LeftBrace, "{", self.line, self.column, self.column)
            case "}":
                tok = Token(TokenType.RightBrace, "}", self.line, self.column, self.column)
            case "[":
                tok = Token(TokenType.LeftBracket, "[", self.line, self.column, self.column)
            case "]":
                tok = Token(TokenType.RightBracket, "]", self.line, self.column, self.column)
            case ":":
                tok = Token(TokenType.Colon, ":", self.line, self.column, self.column)
            case ";":
                tok = Token(TokenType.Semicolon, ";", self.line, self.column, self.column)
            case ",":
                tok = Token(TokenType.Comma, ",", self.line, self.column, self.column)
            case _:
                if self.current_char.isdigit():
                    st_col = self.column
                    num, float_ = self._read_num()
                    if float_:
                        tok = Token(TokenType.Float, num, self.line, st_col, self.column)
                    else:
                        tok = Token(TokenType.Int, num, self.line, st_col, self.column)
                    return tok
                elif self.current_char.isalpha() or self.current_char == "_":
                    st_col = self.column
                    ident = self.current_char
                    self._read_char()
                    while self.current_char.isalnum() or self.current_char == "_":
                        ident += self.current_char
                        self._read_char()
                    match ident:
                        case "is":
                            tok = Token(TokenType.Is, ident, self.line, st_col, self.column - 1)
                        case "and":
                            tok = Token(TokenType.And, ident, self.line, st_col, self.column - 1)
                        case "or":
                            tok = Token(TokenType.Or, ident, self.line, st_col, self.column - 1)
                        case "not":
                            tok = Token(TokenType.Not, ident, self.line, st_col, self.column - 1)
                        case "def":
                            tok = Token(TokenType.Def, ident, self.line, st_col, self.column - 1)
                        case "return":
                            tok = Token(TokenType.Return, ident, self.line, st_col, self.column - 1)
                        case "if":
                            tok = Token(TokenType.If, ident, self.line, st_col, self.column - 1)
                        case "else":
                            tok = Token(TokenType.Else, ident, self.line, st_col, self.column - 1)
                        case "elIf":
                            tok = Token(TokenType.ElIf, ident, self.line, st_col, self.column - 1)
                        case "while":
                            tok = Token(TokenType.While, ident, self.line, st_col, self.column - 1)
                        case "for":
                            tok = Token(TokenType.For, ident, self.line, st_col, self.column - 1)
                        case "in":
                            tok = Token(TokenType.In, ident, self.line, st_col, self.column - 1)
                        case "break":
                            tok = Token(TokenType.Break, ident, self.line, st_col, self.column - 1)
                        case "continue":
                            tok = Token(TokenType.Continue, ident, self.line, st_col, self.column - 1)
                        case "class":
                            tok = Token(TokenType.Class, ident, self.line, st_col, self.column - 1)
                        case "true":
                            tok = Token(TokenType.True_, ident, self.line, st_col, self.column - 1)
                        case "false":
                            tok = Token(TokenType.False_, ident, self.line, st_col, self.column - 1)
                        case "none":
                            tok = Token(TokenType.None_, ident, self.line, st_col, self.column - 1)
                        case _:
                            tok = Token(TokenType.Identifier, ident, self.line, st_col, self.column - 1)
                    return tok
                elif self.current_char == "\"":
                    st_col = self.column
                    self._read_char()
                    string = ""
                    while self.current_char != "\"":
                        string += self.current_char
                        self._read_char()
                    tok = Token(TokenType.String, string, self.line, st_col, self.column)
                elif self.current_char == "'":
                    st_col = self.column
                    self._read_char()
                    raw_string = ""
                    while self.current_char != "'":
                        raw_string += self.current_char
                        self._read_char()
                    tok = Token(TokenType.RawString, raw_string, self.line, st_col, self.column)
                else:
                    tok = Token(TokenType.ILLEGAL, self.current_char, self.line, self.column, self.column)
        self._read_char()
        return tok
    
    def _read_num(self) -> tuple[str, bool]:
        point = False
        num = ""
        while (self.current_char.isdigit() or self.current_char == ".") and self.current_char != "":
            if self.current_char == ".":
                point = True
            num += self.current_char
            self._read_char()
        return num, point

    def _read_char(self) -> None:
        if self.pos >= len(self.source):
            self.current_char = ""
        else:
            self.current_char = self.source[self.pos]
        self.pos += 1
        self.column += 1

    def _peek_char(self, n: int = 1) -> str:
        if self.pos + n >= len(self.source):
            return ""
        return self.source[self.pos - 1 + n]

    def _skip_blanks(self) -> None:
        while self.current_char.isspace():
            if self.current_char == "\n":
                self.line += 1
                self.column = 0
            self._read_char()


def test_all_possible_tokens():
    from rich import print
    # Sample input text containing all possible tokens
    input_text = """
if (x == 10) {
    print("Hello, World!");
} else {
    x = x + 1;
}
while (y != 20) {
    y -= 2;
}
for (int i = 0; i < 10; i++) {
    continue;
}
return x / y;
x += 1;
x -= 1;
x *= 1;
x /= 1;
x %= 1;
x \\= 1;
x && y;
x || y;
x ^ y;
~x;
x << 1;
x >> 1;
x++;
x--;
x.is(y);
x.in(y);
def func():
    pass
class MyClass:
    pass
True
False
None
"""
    
    # Expected tokens (adjust based on your lexer implementation)
    expected_tokens = [
    ]
    
    # Initialize the lexer
    lexer = Lexer(input_text)
    
    # Tokenize the input text
    tokens = lexer.tokens
    print(tokens)
    # Assert the tokens
    # assert tokens == expected_tokens

if __name__ == "__main__":
    test_all_possible_tokens()